### ISSUE TITLE
Race Condition results in occasional slow client response times.

### ISSUE THREAD

**Issue by billyyarosh** [ReactiveX/RxNetty#432]
We are experiencing delays in client requests from time to time when working with RxNetty HttpClient.

 We traced it down and If you do a `toBlocking()` it happens more often, but we think it still happens in a pure async, but less often. That is harder to confirm since we had to add pauses between each request in the main thread.

**NioEventLoop selector.select does not get interrupted** if the write request task happens at the same.

Groovy code we used which resolved the occasional slow requests.

``` groovy

import io.netty.util.concurrent.SingleThreadEventExecutor

import java.lang.reflect.Field
import java.lang.reflect.Modifier
import java.util.concurrent.TimeUnit

Field field = SingleThreadEventExecutor.class.getDeclaredField("SCHEDULE_PURGE_INTERVAL")
field.accessible = true
field.modifiers = Modifier.PUBLIC | Modifier.STATIC
field.set(null, TimeUnit.MILLISECONDS.toNanos(100))
```

I don't know enough about the Netty or the RXNetty to know the resolution, just that there seems to be some race condition occurring.

I'd like for someone in the #382 thread to confirm the above yields consistent load times for them.

**Comment by NiteshKant** [comment]
> > > NioEventLoop selector.select does not get interrupted if the write request task happens at the same.

The selector only gets woken up on `flush()` and not on every write.

If you can show the code that reproduces this problem, it will be useful, else it is very difficult to debug such issues.

**Comment by billyyarosh** [comment]
Here is a groovy script that shows the top five slowest results in milliseconds. In this example I only get a couple blocked requests. 

My results from last run:

```
RxNetty Average: 2.4614614615
RxNetty Top Five Slowest Responses (ms): [863, 791, 472, 25, 5]
```

``` groovy
import io.netty.buffer.ByteBuf
import io.netty.handler.codec.http.HttpResponseStatus
import io.netty.util.concurrent.SingleThreadEventExecutor
import io.reactivex.netty.RxNetty
import io.reactivex.netty.contexts.RxContexts
import io.reactivex.netty.protocol.http.client.HttpClientBuilder
import io.reactivex.netty.protocol.http.client.HttpClientRequest
import io.reactivex.netty.protocol.http.server.HttpServer
import io.reactivex.netty.protocol.http.server.RequestHandler
import rx.functions.Func1

import java.lang.reflect.Field
import java.lang.reflect.Modifier
import java.util.concurrent.TimeUnit
import java.util.function.Function

class RxNettyMain {

    static void main(String[] args) throws URISyntaxException {

        def baseUri = new URL("http://127.0.0.1:1337")
        def relativePath = "/"

        HttpServer<ByteBuf, ByteBuf> server = RxNetty.createHttpServer(1337, {request, response ->
                    response.setStatus(HttpResponseStatus.OK);
                    response.writeString("OK");
                    return response.close();
                } as RequestHandler)
        server.start();

        HttpClientBuilder<ByteBuf, ByteBuf> clientBuilder
        clientBuilder = RxContexts.<ByteBuf, ByteBuf> newHttpClientBuilder(baseUri.getHost(), baseUri.getPort(),
                                                                           RxContexts.DEFAULT_CORRELATOR)

        def rxclient = clientBuilder.build()

        // What we used to hack quicker response times.
//        Field field = SingleThreadEventExecutor.class.getDeclaredField("SCHEDULE_PURGE_INTERVAL")
//        field.accessible = true
//        field.modifiers = Modifier.PUBLIC | Modifier.STATIC
//        field.set(null, TimeUnit.MILLISECONDS.toNanos(100))


        def times = (1..1000).collect {
            timeIt {
                def result = rxclient
                        .submit(HttpClientRequest.createGet(relativePath))
                        .flatMap { contentHolder ->
                            assert contentHolder.status.code() == 200
                            return contentHolder.getContent();
                        }
                        .map(new Func1<ByteBuf, ByteBuf>() {
                            ByteBuf call(ByteBuf o) {
                                return o.copy()
                            }
                        })
                        .toBlocking()
                        .singleOrDefault(null);
            }
        }.tail()



        println "RxNetty Average: ${times.sum() / times.size()}"
        println "RxNetty Top Five Slowest Responses (ms): ${times.sort(false).reverse().take(5)}"
    }

    static long timeIt(Function f) {
        def start = System.nanoTime()
        f.apply(null);
        def end = System.nanoTime()

        return TimeUnit.NANOSECONDS.toMillis(end - start)
    }
}
```

Currently tested on `jdk1.8.0_45`

**Comment by billyyarosh** [comment]
Just to verify it has something to do with the purge interval, remove the .tail() call and run the request twice `(1..2)`.

Your array results will always return: `~ [858, 247]` which is nearly 1 second and is the length of the cleaner purge interval. Increase the purge interval and you will see the two request times increase.

**Comment by NiteshKant** [comment]
@keaplogik I think this is the same behavior as described in issue #412 (for branch 0.5.x). I wil fix it in both branches. Thanks for reporting and apologies for the delayed response.

**Comment by billyyarosh** [comment]
That sounds correct and thanks.

**Comment by dfjones** [comment]
I believe I am also running into this issue (in prod and test code). I'm wondering if this workaround that @keaplogik posted is safe to use in production?

```
Field field = SingleThreadEventExecutor.class.getDeclaredField("SCHEDULE_PURGE_INTERVAL")
field.accessible = true
field.modifiers = Modifier.PUBLIC | Modifier.STATIC
field.set(null, TimeUnit.MILLISECONDS.toNanos(100))
```

Are there any potential downsides to this at low RxNetty request activity?

**Comment by NiteshKant** [comment]
@dfjones the downside will be that the purge task will run too often (once every 100 nanos), the overhead of which depends on how many tasks are cancelled at any time.

I am hoping to release a fix by next week, if you can wait that long.

**Comment by NiteshKant** [comment]
Fixed via #435 

The sample code in java (copied from the PR) is below:

``` java

        HttpServer<ByteBuf, ByteBuf> server = RxNetty.createHttpServer(0, (request, response) -> {
            response.setStatus(HttpResponseStatus.OK);
            response.writeString("OK");
            return response.close();
        });

        server.start();

        HttpClientBuilder<ByteBuf, ByteBuf> clientBuilder =
                RxNetty.<ByteBuf, ByteBuf>newHttpClientBuilder("localhost", server.getServerPort());

        HttpClient<ByteBuf, ByteBuf> rxclient = clientBuilder.build();

        List<Long> durations = new ArrayList<>();

        Observable.range(1, 1000)
                  .toBlocking()
                  .forEach(integer -> {
                      long start = System.currentTimeMillis();
                      rxclient.submit(HttpClientRequest.createGet("/"))
                              .flatMap(response -> {
                                  assert response.getStatus().code() == 200;
                                  return response.getContent();
                              })
                              .toBlocking()
                              .singleOrDefault(null);
                      long end = System.currentTimeMillis();
                      durations.add(end - start);
                  });

        System.out.println("Count distinct durations (ms):  ");

        Observable.from(durations)
                  .toSortedList()
                  .flatMap(Observable::from)
                  .groupBy(aLong -> aLong)
                  .<String>flatMap(go -> go.count()
                                           .map(count -> String.valueOf(go.getKey()) + " : " + count))
                  .toBlocking()
                  .forEach(System.out::println);

```
